src/suggestion-box.ts

```diff
@@ -0,0 +1,409 @@
+            (this.input_el as HTMLInputElement).value =
+                this.result[indextarget][this.listkey];
+            this.show_container = false;
+        } else if (indextarget === this.result.length) {
+            this._loadmoreEntry(event);
+        }
+    }
+    private _mouseOverEvent(e: any) {
+        let curr_index = 0;
+        curr_index = this._indexOfTarget(e);
+        this.index = curr_index;
+        if (this.index >= 0) {
+            const prevItem = this.private_items[this.index];
+            if (prevItem) {
+                prevItem.classList.remove("sgn-active");
+            }
+        }
+        if (curr_index >= 0) {
+            this.private_items[this.index].classList.add("sgn-active");
+        }
+        this.prevIndex = curr_index;
+    }
+    private _inputSelectEvent() {
+        this.index = -1;
+    }
+    private _inputFocusEvent() {
+        if (this.show_empty) this.show_container = true;
+        this.index = -1;
+    }
+    private _mouseOutEvent() {
+        if (this.prevIndex >= 0 && this.prevIndex < this.result.length) {
+            this.private_items[this.index]?.classList.remove("sgn-active");
+        }
+    }
+    windowClickEvent(event: any) {
+        let el = event.target;
+        while (el) {
+            if (el == this) return;
+            el = el.parentElement;
+        }
+        this.show_container = false;
+    }
+    private _format(str: any, obj: any, keys: string) {
+        const input = (this.input_el as HTMLInputElement)?.value;
+        if (typeof obj == "string" || typeof obj == "number") return obj;
+        else {
+            for (let key of Object.keys(obj)) {
+                let template: any;
+                if (key === keys) {
+                    template =
+                        this.highlight === "matched"
+                            ? this._highlight(obj[key], input.toLowerCase())
+                            : this._startsWithHighlight(obj[key]);
+                } else {
+                    template = obj[key];
+                }
+
+                str = str.replace(
+                    new RegExp("\\{" + key + "\\}", "gi"),
+                    template
+                );
+            }
+            return str;
+        }
+    }
+    keyDownInputEvent(e: any) {
+        const resultLength = this.result.length;
+        switch (e.key) {
+            case "ArrowUp":
+                this.index =
+                    this.index == -1
+                        ? resultLength - 1 + (this.has_more ? 1 : 0)
+                        : this.index - 1;
+
+                this._setActive(resultLength);
+                break;
+            case "ArrowDown":
+                this.index =
+                    this.index === -1 ||
+                    (this.has_more && this.index === resultLength)
+                        ? 0
+                        : (this.index + 1) %
+                          (resultLength + (this.has_more ? 1 : 0));
```

Comment:
I think the above condition can be written as 
`this.index = (this.index + 1) % (resultLength + (this.has_more?1:0))`